src/protagonist/Portal/Features/Batches/Requests/ParseCsv.cs

```diff
@@ -0,0 +1,198 @@
+                    if (distinctRows.ContainsKey(record.Line.Value)) continue;
+                    distinctRows.Add(record.Line.Value, new Image
+                    {
+                        CustomerId = customerId,
+                        ModelId = record.Id,
+                        Space = record.Space,
+                        String1 = record.String1,
+                        String2 = record.String2,
+                        String3 = record.String3,
+                        Origin = record.Origin,
+                        Tags = record.Tags.Split(",",
+                            StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries),
+                        Roles = record.Roles.Split(",",
+                            StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries),
+                        MaxUnauthorised = record.MaxUnauthorized ?? -1,
+                        Number1 = record.Number1,
+                        Number2 = record.Number2,
+                        Number3 = record.Number3,
+                        InitialOrigin = record.InitialOrigin,
+                        Family = AssetFamily.Image,
+                        MediaType = "image/jp2"
+                    });
+                }
+                catch (BadDataException badDataEx)
+                {
+                    return ParseCsvResult.Failure(
+                        $"CSV read error: bad data found in file at line {badDataEx.Context.Parser.Row}, row {badDataEx.Context.Reader.CurrentIndex}");
+                }
+                catch (MissingFieldException missingFieldEx)
+                {
+                    return ParseCsvResult.Failure(
+                        $"CSV read error: line {missingFieldEx.Context.Parser.Row} in file is missing fields");
+                }
+                catch (TypeConverterException typeEx)
+                {
+                    var fieldIndex = typeEx.Context.Reader.CurrentIndex - 1;
+                    var currentLine = csv.GetField(1);
+                    readErrors.Add($"Line {currentLine}: could not parse {ImageIngestModel.FieldNames[fieldIndex]} value ({typeEx.Text})");
+                }
+                catch (Exception ex)
+                {
+                    logger.LogError(ex, "Error parsing CSV file");
+                    return ParseCsvResult.Failure("An error occured while parsing the CSV file");
+                }
+            }
+        }
+        
+        // Return an error if parsing fails
+        if (readErrors.Any()) 
+        {
+            return ParseCsvResult.Failure(readErrors.ToArray());
+        }
+        
+        // Order images
+        var images = distinctRows
+            .OrderBy(image => image.Key)
+            .Select(image => image.Value)
+            .ToList();
+        
+        // Split the read images into multiple batches if they surpass the maximum batch size
+        var batches = images.Chunk(maxBatchSize).ToList();
```

Comment:
Internally `.Chunk()` uses [`yield return`](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/yield) to return next element from the iterator. By calling `.ToList()` you are getting all results at once but looking at the code in the `for` statement you might only use the first chunk (e.g. if the first chunk fails then processing is aborted).

You can avoid this by iterating over the `.Chunk()` results, which means you'll only do the work to get the next Chunk of items when they are required. If you abort after chunk 1 of 10, you've only fetched the first rather than all prematurely.

```cs
foreach (var batch in images.Chunk(maxBatchSize))
{
  var collection = new HydraCollection<Image> { Members = batch };
```